pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,292 @@
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
+
+	if client != nil && client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("ttl_after_finished_controller", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	tc := &TTLAfterFinishedController{
+		client:        client,
+		recorder:      eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "ttl-after-finished-controller"}),
+		jobsToCleanup: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttl_jobs_to_delete"),
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    tc.addJob,
+		UpdateFunc: tc.updateJob,
+	})
+
+	tc.jLister = jobInformer.Lister()
+	tc.jListerSynced = jobInformer.Informer().HasSynced
+
+	tc.clock = clock.RealClock{}
+
+	return tc
+}
+
+// Run starts the workers to clean up Jobs.
+func (tc *TTLAfterFinishedController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	glog.Infof("Starting TTL after finished controller")
+	defer glog.Infof("Shutting down TTL after finished controller")
+
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.runJobCleanupWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (tc *TTLAfterFinishedController) addJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Adding job %s", job.Name)
+
+	// Enqueue the Job when TTL expires
+	addJobToCleanupFn := func(j *batch.Job) error {
+		tc.addJobToCleanup(j)
+		return nil
+	}
+	err := tc.checkTTLAndReact(job, addJobToCleanupFn)
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %v: %v", job.Name, err)
+	}
+}
+
+func (tc *TTLAfterFinishedController) updateJob(old, cur interface{}) {
+	job := cur.(*batch.Job)
+	glog.V(4).Infof("Updating job %s", job.Name)
+
+	// Enqueue the Job when TTL expires
+	addJobToCleanupFn := func(j *batch.Job) error {
+		tc.addJobToCleanup(j)
+		return nil
+	}
+	err := tc.checkTTLAndReact(job, addJobToCleanupFn)
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %v: %v", job.Name, err)
+	}
+}
+
+func (tc *TTLAfterFinishedController) addJobToCleanup(job *batch.Job) {
+	glog.V(4).Infof("Add job %s to cleanup", job.Name)
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.jobsToCleanup.Add(key)
+}
+
+func (tc *TTLAfterFinishedController) addJobToCleanupAfter(job *batch.Job, after time.Duration) {
+	glog.V(4).Infof("Add job %s to cleanup after %s", job.Name, after.String())
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.jobsToCleanup.AddAfter(key, after)
```

Comment:
We may want to watch out. This will become a blocking call once there are a thousand jobs who have yet to reach after. For small values of after I do not expect that to be a problem but we may want to consider what happens for large values of after and what thread is being blocked.

Reply:
Added a TODO for now.

Reply:
@cheftako code updated to use a heap to keep track of those jobs, and remove the use of AddAfter